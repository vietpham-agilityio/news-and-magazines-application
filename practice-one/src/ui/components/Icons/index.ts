export * from '@/ui/components/Icons/User';
export * from '@/ui/components/Icons/Plus';
export * from '@/ui/components/Icons/Home';
export * from '@/ui/components/Icons/File';
export * from '@/ui/components/Icons/Close';
export * from '@/ui/components/Icons/Circle';
export * from '@/ui/components/Icons/Search';
export * from '@/ui/components/Icons/SaveTag';
export * from '@/ui/components/Icons/MailBox';
export * from '@/ui/components/Icons/Twitter';
export * from '@/ui/components/Icons/ChatBox';
export * from '@/ui/components/Icons/Ellipse';
export * from '@/ui/components/Icons/Schedule';
export * from '@/ui/components/Icons/ArrowTop';
export * from '@/ui/components/Icons/ThreeDots';
export * from '@/ui/components/Icons/Triangle';
export * from '@/ui/components/Icons/Instagram';
export * from '@/ui/components/Icons/BurgerMenu';
export * from '@/ui/components/Icons/ChevronsUp';
export * from '@/ui/components/Icons/ArrowLeft';
export * from '@/ui/components/Icons/ArrowBack';
export * from '@/ui/components/Icons/ArrowRight';
export * from '@/ui/components/Icons/PaperPlane';
export * from '@/ui/components/Icons/ArrowBottom';
export * from '@/ui/components/Icons/ChatBoxSmall';
export * from '@/ui/components/Icons/SaveTagSmall';