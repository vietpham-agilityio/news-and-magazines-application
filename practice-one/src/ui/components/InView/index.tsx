'use client';

import { InView } from 'react-intersection-observer';

interface IProps {
  children: React.ReactNode;
}

const InViewWrapper = ({ children }: IProps) => {
  return (
    <InView triggerOnce threshold={1}>
      {({ ref, inView }) => (
        <div ref={ref} className="w-full">
          {inView && children}
        </div>
      )}
    </InView>
  );
};

export default InViewWrapper;
