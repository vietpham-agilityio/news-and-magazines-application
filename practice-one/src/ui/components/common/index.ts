export * from '@/ui/components/common/Tag';
export * from '@/ui/components/common/Menu';
export * from '@/ui/components/common/Title';
export * from '@/ui/components/common/Input';
export * from '@/ui/components/common/Avatar';
export * from '@/ui/components/common/Button';
export * from '@/ui/components/common/BoxIcon';
export * from '@/ui/components/common/SubMenu';
export * from '@/ui/components/common/TextArea';
export * from '@/ui/components/common/Typography';
export * from '@/ui/components/common/Information';
