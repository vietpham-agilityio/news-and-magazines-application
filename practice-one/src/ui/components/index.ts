// common
export * from '@/ui/components/common';

// shared
export * from '@/ui/components/Brand';
export * from '@/ui/components/NavItem';
export * from '@/ui/components/Comment';
export * from '@/ui/components/Carousel';
export * from '@/ui/components/SearchBar';
export * from '@/ui/components/PaperPost';
export * from '@/ui/components/BurgerMenu';
export * from '@/ui/components/Pagination';
export * from '@/ui/components/CommentForm';
export * from '@/ui/components/BreadCrumbs';
export * from '@/ui/components/TitleSection';
export * from '@/ui/components/SubmitButton';
export * from '@/ui/components/Card/CardSingle';
export * from '@/ui/components/Skeleton/Comment';
export * from '@/ui/components/Skeleton/CardPost';
export * from '@/ui/components/Skeleton/BadgeGroup';
export * from '@/ui/components/Skeleton/DetailPost';
export * from '@/ui/components/Skeleton/CardVertical';
export * from '@/ui/components/Skeleton/CardCategory';
export { default as CardPost } from '@/ui/components/Card/CardPost';
export { default as CardCategory } from '@/ui/components/Card/CardCategory';