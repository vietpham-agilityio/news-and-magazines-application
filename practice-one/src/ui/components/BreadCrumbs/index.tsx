// components
import Link from 'next/link';
import { Typography } from '@/ui/components';

// types
import { FontWeight, IBreadCrumbItem, Size } from '@/types';

// icons
import { ArrowRight } from '@/ui/components/Icons';

interface IProps {
  listBreadCrumb: IBreadCrumbItem[];
}

export const BreadCrumbs = ({ listBreadCrumb }: IProps) => (
  <div className="flex items-center">
    {listBreadCrumb.map((item, index) => {
      const { path, name } = item;

      return (
        <div key={name} className="flex items-center">
          <Link href={path}>
            <Typography
              tag="h2"
              textSize={Size.XS}
              weight={FontWeight.Normal}
              additionalClasses={`
              ${
                index === listBreadCrumb.length - 1
                  ? 'w-57.5 text-dark-90 h-5 truncate text-ellipsis overflow-hidden sm:w-full'
                  : ' text-dark-100 hover:text-primary-100'
              }`}
            >
              {name}
            </Typography>
          </Link>
          {index !== listBreadCrumb.length - 1 && (
            <i className="mx-2">
              <ArrowRight />
            </i>
          )}
        </div>
      );
    })}
  </div>
);
