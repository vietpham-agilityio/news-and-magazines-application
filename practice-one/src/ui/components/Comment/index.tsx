// static image
import ImageStore from '@/constants/images';

// components
import { Avatar, Button, Typography } from '@/ui/components';

// icons
import { ArrowBack, Schedule } from '@/ui/components/Icons';

// types
import { FontWeight, Size } from '@/types';

interface IProps {
  userName: string;
  content: string;
  publicDate?: string;
}

export const Comment = ({ userName, content, publicDate }: IProps) => (
  <div className="bg-white-90 p-4 rounded-xl">
    { publicDate ? (
      <div className="flex flex-1 px-1 mb-4">
        <Avatar
          width={60}
          height={60}
          url={ImageStore.AvatarUser}
          alt="Author this comment"
        />
        <div className="flex flex-1 items-center justify-between">
          <div className="flex flex-1 justify-between">
            <div className="flex flex-col h-full ml-2.5">
              <Typography
                tag="h2"
                textSize={Size.SM}
                additionalClasses="text-dark-100"
              >
                {userName}
              </Typography>
              <div className="flex gap-2">
                <Schedule />
                <Typography
                  tag="h3"
                  weight={FontWeight.Medium}
                  additionalClasses="text-dark-75 text-p leading-5"
                >
                  {publicDate}
                </Typography>
              </div>
            </div>
          </div>
        </div>
      </div>
    ) : (
      <Typography tag="h2" textSize={Size.SM} additionalClasses="text-dark-100">
        {userName}
      </Typography>
    )}
    <Typography
      tag="h3"
      weight={FontWeight.Light}
      additionalClasses={`${publicDate ? 'text-xsm text-black-75 leading-5' : 'text-p text-dark-75'}`}
    >
      {content}
    </Typography>
  </div>
);
