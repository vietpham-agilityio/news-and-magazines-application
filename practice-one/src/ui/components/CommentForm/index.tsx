// components
import { createComment } from '@/services';
import { Input, SubmitButton, TextArea } from '@/ui/components';

// icons
import { ChatBoxSmall } from '@/ui/components/Icons';

interface IProps {
  postId: string;
}

export const CommentForm = ({postId}: IProps) => {

return (
  <form action={createComment}>
    <div className="grid grid-cols-12 grid-rows-3 lg:gap-6">
      <div className="row-span-3 col-span-12 lg:col-span-6">
        <input name="postId" type="hidden" value={postId} />
        <div className="mb-5 row-span-1">
          <Input label="Name" name="name" type="text" />
        </div>
        <div className="mb-5 row-span-1">
          <Input label="Email" name="email" type="email" />
        </div>
        <div className="mb-5 row-span-1">
          <Input label="Website" name="website" type="text" />
        </div>
      </div>
      <div className="row-span-3 col-span-12 lg:col-span-6">
        <TextArea name='content' label="Comment" placeholder="What do you think?" />
      </div>
    </div>
    <div className="flex w-full justify-end mt-7 mb-30 sm:mt-10 sm:mb-12 lg:mt-0">
      <SubmitButton
        label="Send Comment"
        leftIcon={<ChatBoxSmall />}
      />
    </div>
  </form>
)}
