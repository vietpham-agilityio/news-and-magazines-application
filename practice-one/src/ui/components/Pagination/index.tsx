'use client';

import { useState } from 'react';

// components
import Link from 'next/link';
import { useRouter } from 'next/navigation';
import { Button, Typography } from '@/ui/components';

// icons
import { ArrowLeft, ArrowRight } from '@/ui/components/Icons';

// types
import { Size } from '@/types';

export interface IProps {
  pageCount: number;
  categoryId: number;
}

export const Pagination = ({ pageCount, categoryId }: IProps) => {
  const router = useRouter();
  const [pageIndex, setPageIndex] = useState<number>(1);

  const handlePrevious = () => {
    const newPageIndex = pageIndex - 1;
    setPageIndex(newPageIndex);
    router.push(`/category/${categoryId}/?page=${newPageIndex}`);
  };

  const handleNext = () => {
    const newPageIndex = pageIndex + 1;
    setPageIndex(newPageIndex);
    router.push(`/category/${categoryId}/?page=${newPageIndex}`);
  };

  const handlePagination = (pageIndex: number) => {
    setPageIndex(pageIndex);
  };

  return (
    <div className="h-10 mt-10 mb-30 sm:mt-11 sm:mb-12 lg:mb-25 flex items-center">
      <div className="mx-auto flex items-center justify-between px-4 py-3 sm:mx-0 sm:px-6">
        <div className="flex sm:flex-1 sm:items-center sm:justify-between">
          <div className="w-90">
            <nav className="flex gap-1" aria-label="Pagination">
              {pageIndex !== 1 && (
                <Button
                  dataTestId="previous-pagination-button"
                  isMarginBottom
                  label="Previous"
                  textColor="text-dark-75"
                  bgColor="bg-white-90"
                  leftIcon={<ArrowLeft color="#404040" />}
                  onClick={handlePrevious}
                />
              )}
              {[...Array(pageCount)].map((_, index) => {
                const pageNumber = index + 1;
                return (
                  <Link
                    key={pageNumber}
                    href={`/category/${categoryId}/?page=${pageNumber}`}
                    data-testid={`pagination-button-${pageNumber}`}
                    aria-label={`pagination-button-${pageNumber}`}
                    className={`w-10 h-10 flex items-center justify-center rounded-xl hover:bg-white-90 ${
                      pageIndex === pageNumber && 'bg-white-90'
                    }`}
                    onClick={() => handlePagination(pageNumber)}
                  >
                    <Typography
                      tag="h2"
                      textSize={Size.SM}
                      additionalClasses="text-dark-100"
                    >
                      {pageNumber}
                    </Typography>
                  </Link>
                );
              })}
              {pageIndex !== pageCount && (
                <Button
                  dataTestId="next-pagination-button"
                  isMarginBottom
                  label="Next"
                  textColor="text-dark-75"
                  bgColor="bg-white-90"
                  rightIcon={<ArrowRight color="#404040" />}
                  onClick={handleNext}
                />
              )}
            </nav>
          </div>
        </div>
      </div>
    </div>
  );
};
