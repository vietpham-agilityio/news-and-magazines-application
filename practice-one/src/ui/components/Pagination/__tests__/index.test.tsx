import { fireEvent, render } from '@testing-library/react';

// component
import { Pagination } from '@/ui/components';
import { useRouter } from 'next/navigation';

jest.mock('next/navigation', () => {
  const originalModule = jest.requireActual('next/navigation');

  return {
    ...originalModule,
    useRouter: jest.fn(),
  };
});

describe('Pagination', () => {
  const mockUseRouter = useRouter as jest.Mock;
  mockUseRouter.mockReturnValue({
    push: jest.fn(),
  });

  const renderPagination = () => render(<Pagination categoryId={2} pageCount={4} />);

  it('should render Pagination macth snapshot', () => {
    const { asFragment } = renderPagination();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Pagination should trigger event when clicked', () => {
    const { getByTestId, getByText } = renderPagination();

    const secondaryButton = getByTestId('pagination-button-2');

    expect(secondaryButton).toBeInTheDocument();

    fireEvent.click(secondaryButton);

    const previousButton = getByText('Previous');
    const nextButton = getByText('Next');

    expect(previousButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
  });

  it('Pagination should render ui when trigger event', () => {
    const { getByTestId } = renderPagination();

    const secondaryButton = getByTestId('pagination-button-2');

    expect(secondaryButton).toBeInTheDocument();

    fireEvent.click(secondaryButton);

    const previousButton = getByTestId('previous-pagination-button');

    expect(previousButton).toBeInTheDocument();

    fireEvent.click(previousButton);

    const nextButton = getByTestId('next-pagination-button');

    expect(nextButton).toBeInTheDocument();

    fireEvent.click(nextButton);
  });
});
