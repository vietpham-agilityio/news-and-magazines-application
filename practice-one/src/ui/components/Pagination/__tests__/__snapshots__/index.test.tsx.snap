// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render Pagination macth snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-10 mt-10 mb-30 sm:mt-11 sm:mb-12 lg:mb-25 flex items-center"
  >
    <div
      class="mx-auto flex items-center justify-between px-4 py-3 sm:mx-0 sm:px-6"
    >
      <div
        class="flex sm:flex-1 sm:items-center sm:justify-between"
      >
        <div
          class="w-90"
        >
          <nav
            aria-label="Pagination"
            class="flex gap-1"
          >
            <a
              aria-label="pagination-button-1"
              class="w-10 h-10 flex items-center justify-center rounded-xl hover:bg-white-90 bg-white-90"
              data-testid="pagination-button-1"
              href="/category/2?page=1"
            >
              <h2
                aria-level="2"
                class="text-sm font-medium text-dark-100"
                role="heading"
              >
                1
              </h2>
            </a>
            <a
              aria-label="pagination-button-2"
              class="w-10 h-10 flex items-center justify-center rounded-xl hover:bg-white-90 false"
              data-testid="pagination-button-2"
              href="/category/2?page=2"
            >
              <h2
                aria-level="2"
                class="text-sm font-medium text-dark-100"
                role="heading"
              >
                2
              </h2>
            </a>
            <a
              aria-label="pagination-button-3"
              class="w-10 h-10 flex items-center justify-center rounded-xl hover:bg-white-90 false"
              data-testid="pagination-button-3"
              href="/category/2?page=3"
            >
              <h2
                aria-level="2"
                class="text-sm font-medium text-dark-100"
                role="heading"
              >
                3
              </h2>
            </a>
            <a
              aria-label="pagination-button-4"
              class="w-10 h-10 flex items-center justify-center rounded-xl hover:bg-white-90 false"
              data-testid="pagination-button-4"
              href="/category/2?page=4"
            >
              <h2
                aria-level="2"
                class="text-sm font-medium text-dark-100"
                role="heading"
              >
                4
              </h2>
            </a>
            <button
              aria-label="button-element"
              class="flex items-center rounded-xl bg-white-90 py-2.5 undefined pr-4 pl-6 hover:opacity-80"
              data-testid="next-pagination-button"
            >
              <h2
                aria-level="2"
                class="text-xsm font-medium text-dark-75 leading-5"
                role="heading"
              >
                Next
              </h2>
              <i
                class="ml-2"
              >
                <svg
                  data-testid="arrow-right"
                  fill="none"
                  height="11"
                  width="7"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.25 5.5a.866.866 0 0 1-.246.629l-4.375 4.375a.843.843 0 0 1-1.23 0 .843.843 0 0 1 0-1.23L4.144 5.5.398 1.754a.843.843 0 0 1 0-1.23.843.843 0 0 1 1.23 0l4.376 4.374a.848.848 0 0 1 .246.602Z"
                    fill="#404040"
                    fill-opacity="0.5"
                  />
                </svg>
              </i>
            </button>
          </nav>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
